dashboard/lti_new.py

```diff
@@ -75,6 +79,48 @@ def get_jwks(request):
     return JsonResponse(config.get_jwks())
 
 
+def recursive_format(args, **kwargs):
+    """
+    Recursively apply `string.format()` to all strings in a data structure.
+
+    This is intended to be used on a data structure that may contain
+    format strings just before it is passed to `json.dump()` or `dumps()`.
+
+    :param args: data structure; mostly dictionaries and lists
+    :param kwargs: values to be used in `string.format()`
+    :return: a new data structure with formatted strings
+    """
+    if isinstance(args, Mapping):
+        return {key: recursive_format(value, **kwargs)
+                for (key, value) in args.items()}
+    elif isinstance(args, MutableSequence):
+        return [recursive_format(value, **kwargs)
+                for value in args]
+    else:
+        return args.format(**kwargs)
+
+
+def generate_config_json(request: HttpRequest) -> Union[
+    JsonResponse, LTIError]:
+    config = get_tool_conf()
```

Comment:
What is the need for doing this? This is something PyLTI1.3 library after grabbing the LTI_CONFIG values try to do the LTI login/launch. when we configure the `/lti/config/` Myla simply return the JSON.

Reply:
Because the LTI should be enabled and configured properly before an LMS requests the config and tries to use it.

That seems important to me.

Reply:
@pushyamig (or @jonespm), can you confirm whether checking MyLA's LTI config before generating the config for the LMS is OK, or does it cause a problem?

Reply:
I tested it locally and it did create configuration and I pointed those config in my local instance and it worked as expected. I did that yesterday and not with updated changes. Should work even with recent changes as it is not something drastically different

Reply:
@lsloan I don't see any harm done, it just instantiate the tool config. I personally would have not done it. Since this is purely communication between Myla and Canvas to get the LTI keys. Even let's say if we did not put any LTI_CONFIGS then it will throw errors. I don't know if Canvas notifies with the response we are sending or defaults to its own error message. I am fine either way

Reply:
To test what happens when MyLA's LTI configuration is invalid, I altered the code a little to simulate that.

## Results

### Canvas

Canvas showed the message

> Request failed with status code 422

MyLA's response actually has a status code of 500.  I believe 422 is the response from Canvas' internal processing to its UI.  (422 is "Unprocessable Entity", probably the error given when Canvas didn't find the expected content in the LTI config JSON.)

That message is good enough to tell the user something was wrong with MyLA.

### MyLA

MyLA logged an error message.  I'm making some changes to that log message now, to be a little more specific.

> [2020-10-16 14:58:16 -0400] [ERROR] [lti_new.py:72] Invalid LTI configuration: "0"

That message should help MyLA administrators that something is wrong.

The content of the response also includes the error.

```json
{"lti_error": "Invalid LTI configuration: \"0\""}
```

## Conclusion

It's helpful to check the LTI configuration, even if it's not being used in that block of code.

Reply:
Thanks for the detailed text. I guess Canvas will show the error response based on it's interpretation and it is fine that way. +1 with log message